utils/litemap/src/store/vec_impl.rs

```diff
@@ -174,6 +149,76 @@ impl<K, V> StoreIntoIterator<K, V> for Vec<(K, V)> {
     fn lm_extend_start(&mut self, other: Self) {
         self.splice(0..0, other);
     }
+
+    /// Extends this store with items from an iterator.
+    #[inline]
+    fn lm_extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = (K, V)>,
+        K: Ord,
+    {
+        // From HashBrown's implementation:
+        // Keys may be already present or show multiple times in the iterator.
+        // Reserve the entire hint lower bound if the map is empty (e.g. from_iter).
+        // Otherwise reserve half the hint (rounded up), so the map
+        // will only resize twice in the worst case (assuming growth by doubling).
+        let iter = iter.into_iter();
+        let (size_hint_lower, _) = iter.size_hint();
+        let reserve = if self.is_empty() {
+            size_hint_lower
+        } else {
+            (size_hint_lower + 1) / 2
+        };
+        self.reserve_exact(reserve);
```

Comment:
Suggestion:

This function is a bit hard to follow, and I'm not convinced it is always correct, especially the things involving `sorted_len`.

We handle some special cases (such as the append-items already being in order) but not others (such as when they are all less than the target).

How about you just append everything to the end and have a single call to `self.sort_by`, and assume that `self.sort_by` will have a fast path if the things are in order. Easy to review, and the worst case is still `N log N`.

Reply:
> How about you just append everything to the end and have a single call to self.sort_by, .... Easy to review, and the worst case is still N log N.

That was my thinking in 5bab562b5b9b26cc79ba49a1f4439a6428b241da too, but I realized that the required deduplication step (after sort) had quadratic behavior. That became clean when I added `bench_extend_rand_dups`.